### PR TITLE
Fix ForwardDiff Dual detection for circular references

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
Fixes https://github.com/SciML/DiffEqBase.jl/issues/784

### CODE REVIEW

File: src/forwarddiff.jl [hunk 1]
```diff
@@ -91,44 +96,47 @@ Base.@pure function anyeltypedual(::Type{T}) where {T <: Union{NTuple}}
...
-                                                                                  AbstractArray{
-                                                                                                <:Number
-                                                                                                },
-                                                                                  Set{
-                                                                                      <:Number
-                                                                                      },
-                                                                                  NTuple{N,
-                                                                                         <:Number
-                                                                                         }}}
+function anyeltypedual(x::Union{Array{T}, AbstractArray{T}, Set{T}},
```
Comment (YingboMa) [inline src/forwarddiff.jl:110]: That's impressive formatting.

File: src/forwarddiff.jl [hunk 2]
```diff
@@ -57,29 +57,34 @@ and then attempt to upconvert `u0` to match that dual-ness. Because this changes
...
 end
 
 # Opt out since these are using for preallocation, not differentiation
-anyeltypedual(x::Union{ForwardDiff.AbstractConfig, Module}) = Any
-anyeltypedual(x::Type{T}) where {T <: Union{ForwardDiff.AbstractConfig}} = Any
+anyeltypedual(x::Union{ForwardDiff.AbstractConfig, Module}, counter = 0) = Any
+anyeltypedual(x::Type{T}, counter = 0) where {T <: Union{ForwardDiff.AbstractConfig}} = Any
 
-Base.@pure function anyeltypedual(::Type{T}) where {T}
+Base.@pure function anyeltypedual(::Type{T}, counter = 0) where {T}
```
Comment (YingboMa) [inline src/forwarddiff.jl:79]: I think `@pure` annotation on generic function is not legal.

Comment (ChrisRackauckas) [reply src/forwarddiff.jl:79]: This should be safe unless someone mutates `T.parameters` though?

Comment (ChrisRackauckas) [reply src/forwarddiff.jl:79]: Do a follow up if you're interested. I'm not sure how to test that so I'll leave it to you.

**Pull request merged**